packages/SwingSet/docs/devices.md

```diff
@@ -255,6 +255,51 @@ However the inbound message pathway uses `dispatch.invoke(deviceNodeID,
 method, argsCapdata) -> resultCapdata`, and the outbound pathway uses
 `syscall.sendOnly`.
 
+## Raw Devices
+
+An alternate way to write a device is to use the "raw device API". In this
+mode, there is no deviceSlots layer, and no attempt to provide
+object-capability abstractions. Instead, the device code is given a `syscall`
+object, and is expected to provide a `dispatch` object, and everything else
+is left up to the device.
+
+This mode makes it possible to create new device nodes as part of the normal
+API, because the code can learn the raw device ref (dref) of the target
+device node on each inbound invocation, without needing a pre-registered
+table of JavaScript `Object` instances for every export.
```

Comment:
I'm wondering if an alternate model would be to implement devices using deviceSlots but treat the root object of a (more sophisticated kind of) device rather the way we treat the root object of a vat: the contact point for bootstrapping further interactions with whatever services it provides.   As part of the API provided by deviceSlots, devices would have access to a `createDeviceNode` power that would create a device node that could be returned when a method is invoked on the root object. In essence, device node 0 would be qualitatively different: a privileged control device that hands out function-specific device nodes, in contrast to the function-specific device nodes themselves which would provide whatever services the "device" is being created to support.

Obviously this would require a somewhat different model than we have now of what it means to be the client of a device, since this controller vs. service distinction would have to be reflected in whatever vat(s) are using the device.

Reply:
yeah, I think a more unified model (which would fully satisfy the goals of #1346) would probably allow the export of both objects *and* device nodes, and then the root could be a plain object, from which you must ask for device nodes via `E()` calls. maybe.